openedx/core/djangoapps/schedules/resolvers.py

```diff
@@ -1,136 +1,313 @@
+        schedule_day_equals_target_day_filter = {
+            '{}__gte'.format(self.schedule_date_field): target_day,
+            '{}__lt'.format(self.schedule_date_field): target_day + datetime.timedelta(days=1),
+        }
+        schedules = Schedule.objects.select_related(
+            'enrollment__user__profile',
+            'enrollment__course',
+        ).prefetch_related(
+            'enrollment__course__modes'
+        ).filter(
+            Q(enrollment__course__end__isnull=True) | Q(
+                enrollment__course__end__gte=self.current_datetime),
+            enrollment__user__in=users,
+            enrollment__is_active=True,
+            **schedule_day_equals_target_day_filter
+        ).order_by(order_by)
+
+        if self.org_list is not None:
+            if self.exclude_orgs:
+                schedules = schedules.exclude(enrollment__course__org__in=self.org_list)
+            else:
+                schedules = schedules.filter(enrollment__course__org__in=self.org_list)
+
+        if "read_replica" in settings.DATABASES:
+            schedules = schedules.using("read_replica")
+
+        LOG.debug('Query = %r', schedules.query.sql_with_params())
+
+        with function_trace('schedule_query_set_evaluation'):
+            # This will run the query and cache all of the results in memory.
+            num_schedules = len(schedules)
+
+        # This should give us a sense of the volume of data being processed by each task.
+        set_custom_metric('num_schedules', num_schedules)
+
+        return schedules
 
 
 class ScheduleStartResolver(BinnedSchedulesBaseResolver):
     """
     Send a message to all users whose schedule started at ``self.current_date`` + ``day_offset``.
     """
-    async_send_task = recurring_nudge_schedule_bin
+    log_prefix = 'Scheduled Nudge'
+    schedule_date_field = 'start'
     num_bins = RECURRING_NUDGE_NUM_BINS
-    enqueue_config_var = 'enqueue_recurring_nudge'
 
-    def __init__(self, *args, **kwargs):
-        super(ScheduleStartResolver, self).__init__(*args, **kwargs)
-        self.log_prefix = 'Scheduled Nudge'
+    def schedules_for_bin(self):
+
+        schedules = self.get_schedules_with_target_date_by_bin_and_orgs()
+
+        for (user, user_schedules) in groupby(schedules, lambda s: s.enrollment.user):
+            user_schedules = list(user_schedules)
+            course_id_strs = [str(schedule.enrollment.course_id) for schedule in user_schedules]
+
+            first_schedule = user_schedules[0]
+            template_context = get_base_template_context(self.site)
+            template_context.update({
+                'student_name': user.profile.name,
+
+                'course_name': first_schedule.enrollment.course.display_name,
+                'course_url': absolute_url(self.site, reverse('course_root', args=[str(first_schedule.enrollment.course_id)])),
+
+                # This is used by the bulk email optout policy
+                'course_ids': course_id_strs,
```

Comment:
Why do we need to check the optout policy for _all_ the courses?  It looks like we're sending an email for only the 1st course.  So shouldn't we honor optout for just that course?

Reply:
Hrm... Yeah, that's a good question. We're aggregating the email across courses, so the theory was that the policy would filter out based on the set of courses being sent about. Unfortunately, I think in the current state, if the first schedule is the one that the policy disallows, then things are going to break.